### PR TITLE
Recursively defined advanced properties (II)

### PR DESCRIPTION
**Pull request by noseglasses**
This is #41 rebased to develop.

### CODE REVIEW

File: cmake/Platform/Core/BoardPropertiesReader.cmake [hunk 1]
```diff
@@ -58,6 +99,7 @@ function(_GET_BOARD_PROPERTY BOARD_ID PROPERTY_NAME OUTPUT_VAR)
     if (NOT VALUE)
         message(FATAL_ERROR "Board info not found: BoardName='${BOARD_NAME}' BoardCPU='${BOARD_CPU}' PropertyName='${PROPERTY_NAME}'")
     endif()
+    _recursively_replace_properties(${BOARD_ID} VALUE)
```
Comment (MrPointer) [inline cmake/Platform/Core/BoardPropertiesReader.cmake:102]: Does it has to be called each time we attempt to get a board's property?
Correct me if I'm wrong, but from my understanding the line `while("${${PROPERTY_VALUE_VAR}}" MATCHES "[^\$]{([^}]*)}")` in the `_recursively_replace_properties` function will return `False` if there are no recursive properties. If so, there's no problem at all with the call. Am I right?

Comment (MrPointer) [reply cmake/Platform/Core/BoardPropertiesReader.cmake:102]: @noseglasses Could you please answer it so we can progress with the PR?

**Comment by fraillt** [comment]
This recursive stuff doesn't solve anything, I know what are you trying to solve, but it should be solved other way around. 
In order to address this issue, try to think of ways how to parse `platform.txt`, because build process consists of these two files `platform.txt` and `boards.txt`.
It works like this (by default for arduino):
* In `platform.txt` in build recipes section, it uses `{build.extra_flags}` which is empty, unless it is defined/overridden in `boards.txt`.
* In `boards.txt` if it is defined for specific board, it has the following `{build.usb_flags}`, which is defined in `platform.txt`.
* In `platform.txt` this property once again uses fields like `{build.vid}` and `{build.pid}` which are defined in `boards.txt`.

Do you see how closely are related these `platform.txt` and `boards.txt` files?
I understand that you want to make it work for your board, and maybe it is even achievable. If all goes well, then maybe it will still be working for default arduino platform, but not more.
The code will be cluttered with workarounds and specific stuff for your board, that only you will know why it is here.
Currently, this project is hardcoded to work with default arduino platform. I think that there are two ways: 
* leave it as it is *OR* (blue pill)
* make it a general, that (red pill)
  * knows how to parse `platform.txt`.
  * knows how to distinguish and find libraries depending on the board architecture, 
  * knows that architecture can contain multiple cores, and build according to selected board
  * knows that in `boards.txt` a board can be fully customizable, not only `menu.CPU`, but arbitrary number of options.
  * knows that multiple tools can exist, for each architecture.
  * knows many other things that I'm not aware of...

If you're ready to start a new journey towards brighter future, then think about this options instead of hacking for your board. 
A nice place to start is `arduino-builder` file, that does all the magic. And if we're able to mimic its behaviour with the same set of flags, then this project will have **exactly the same functionality as Arduino IDE** itself.

**Comment by noseglasses** [comment]
Thanks for all the information.

What this PR addresses is the problem that Arduino allows build variables to be depending on other build variables.

The usb- and extra-flags handling can eventually be made more sophisticated, I agree. But there is also evolution, not only revolution.

**Comment by MrPointer** [comment]
@fraillt @noseglasses I agree with both of you, but in this case @noseglasses is right:

> there is also evolution, not only revolution.

Introducing code that will parse `platform.txt` and such would be great, in fact it would probably mean that we have been successful in our job to "replace" the default Arduino IDE.
However, we must remember that such changes take time & effort, and until we're able to introduce those changes we must also address some existing issues, so that people can still use this project instead of running away disappointed.
@noseglasses Just answer my question in the review please and this PR will be merged.

**Comment by noseglasses** [comment]
> @noseglasses Just answer my question in the review please and this PR will be merged.
Excuse me, what review?

**Review by MrPointer — COMMENTED** [review]
@noseglasses Oh foolish me, it appears I didn't actually post the review and have kept it pending until now... My mistake.

File: cmake/Platform/Core/BoardPropertiesReader.cmake (hunk 1 above)
Comment (noseglasses) [reply cmake/Platform/Core/BoardPropertiesReader.cmake:102]: Yes, you are right. Only if there are any `{...}` matches found, the body of the while loop executes.

**Review by MrPointer — APPROVED** [review]
Looks good, although this change **IS** temporary until a proper parse to `platform.txt` in conjunction with `boards.txt` will be introduced.

**Pull request merged**